### DISCUSSION TITLE
Unsuccessful Building from source locally

### DISCUSSION THREAD

**Discussion by Raynchowkw** [The-OpenROAD-Project/OpenROAD-flow-scripts#20] [Q&A] [answered]
I change from building by docker to building locally, by using `./build_openroad.sh --local`.
It met problem at the very beginning ([3%]). I guess it is for "**yosys**".
The fatal Error messages are:
```
[  3%] Building kernel/cellaigs.o
[  3%] Building kernel/celledges.o
[  3%] Building kernel/satgen.o
kernel/driver.cc:24:33: fatal error: readline/readline.h: No such file or directory
 #  include <readline/readline.h>
                                 ^
compilation terminated.
kernel/yosys.cc:24:33: fatal error: readline/readline.h: No such file or directory
 #  include <readline/readline.h>
                                 ^
compilation terminated.
```

After I `source setup_env.sh` and `cd flow` and `make`, it says
```
(/usr/bin/time -f "%Eelapsed %PCPU %MmemKB" yosys -c scripts/synth.tcl) 2>&1 | tee ./logs/nangate45/gcd/base/1_1_yosys.log
/usr/bin/time: **cannot run yosys**: No such file or directory
0:00.00elapsed 79%CPU 364memKB
make: *** [results/nangate45/gcd/base/1_1_**yosys**.v] Error 127
```

Are there missing of dependencies? Do I need to install all the dependencies manually if I want to build it locally?
I didn't meet this problem when I used docker to build.
Thanks for any advice and help.

**Comment by maliberty** [comment]
If you are on centos or ubuntu you can install all the dependencies with

sudo etc/DependencyInstaller.sh -dev

**Reply by Raynchowkw** [reply]
Hi Maliberty,
I ran this and it met the error: nothing to do. This may be because I manually ran each line in Dockerfile long time ago in my centos machine. This is the message:
```
$ sudo ./etc/DependencyInstaller.sh -dev
epel-release-latest-7.noarch.rpm                         |  15 kB     00:00     
Examining /var/tmp/yum-root-WuL8ip/epel-release-latest-7.noarch.rpm: epel-release-7-13.noarch
/var/tmp/yum-root-WuL8ip/epel-release-latest-7.noarch.rpm: does not update installed package.
Error: Nothing to do
```
And then I tried `./build_openroad.sh --local` again and it met the same problem again.
```
[  3%] Building kernel/satgen.o
kernel/yosys.cc:24:33: fatal error: readline/readline.h: No such file or directory
 #  include <readline/readline.h>
                                 ^
kernel/driver.cc:24:33: fatal error: readline/readline.h: No such file or directory
 #  include <readline/readline.h>
                                 ^
compilation terminated.
compilation terminated.
```
It is like it's missing the "readline" file in the directory "/OpenROAD-flow-scripts/tools/yosys/kernel". I checked it. There is no such file there.
Could you help me with this problem?
Thanks.

**Reply by Raynchowkw** [reply]
I also tried the build step in the openroad .git repo. It cannnot build successfully, either.
I post the issue under openroad .git [here](https://github.com/The-OpenROAD-Project/OpenROAD/issues/632#issue-885059507).
Thanks for any help.

**Reply by Raynchowkw** [reply]
Could you tell me how to check if I install the dependencies correctly?

**Comment by maliberty** [comment]
It seems like this is a duplicate of OpenRoad/632.  I made a suggestion there already that you said "I will try soon."  Let's just discuss in one issue or the other but not both.

**Reply by Raynchowkw** [reply]
Yes, I will discuss "build openroad" there and keep here for discussion of "build openroad-flow-script".
I want to make this run through: `./build_openroad.sh --local`. If it is isn't the problem of `./etc/DependencyInstaller.sh -dev`, I don't know how to find other problems for the "fatal error: readline/readline.h: No such file or directory".

**Comment by Raynchowkw** [comment]
Hi @maliberty ,
I try the local build again with the correct compiler version by 

source scl_source enable devtoolset-8
source scl_source enable llvm-toolset-7.0 
./build_openroad.sh --local

but I still get Error immediately. It is still the same. It says it is missing readline file, as following:
```
$ ./build_openroad.sh --local
[INFO][FLOW-0000] using local build method. This will create binaries at tools/build/
[  0%] Building kernel/version_014c7e26.o
[  0%] Building kernel/driver.o
kernel/driver.cc:24:12: fatal error: readline/readline.h: No such file or directory
 #  include <readline/readline.h>
            ^~~~~~~~~~~~~~~~~~~~~
compilation terminated.
make: *** [Makefile:688: kernel/driver.o] Error 1
make: *** Waiting for unfinished jobs....
[  1%] Building kernel/register.o
```
Could you help me with this?
Is it related to dependencies installation? Is there a way to 'check for nested dependencies' according to dockerfiles?

**Reply by Raynchowkw** [reply]
I think it's because of dependencies of yosys, so I find it in /tools/yosys_util/dockerfile.
I run this line:
RUN `yum install -y flex tcl tcl-devel libffi-devel git graphviz readline-devel glibc-static wget autoconf zlib-devel`
but I get conflicts:
```
Error: git224-core conflicts with git-1.8.3.1-23.el7_8.x86_64
Error: git224 conflicts with git-1.8.3.1-23.el7_8.x86_64
Error: git224-perl-Git conflicts with perl-Git-1.8.3.1-23.el7_8.noarch
 You could try using --skip-broken to work around the problem
 You could try running: rpm -Va --nofiles --nodigest
```
Do they matter?
Thanks.

**Reply by Raynchowkw** [reply, marked as answer]
I am using --skip-broken to work around these.

**Reply by Raynchowkw** [reply]
I think I have worked around it. I can build the flow to 100% now.
